File: src/main/kotlin/raft/war/binary/parser/w3g/parser/replay/LeaveRecord.java
package raft.war.binary.parser.w3g.parser.replay;

import raft.war.binary.parser.w3g.parser.packed.IRecord;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.Objects;

public class LeaveRecord implements IRecord {
    public static final int TYPE = 0x17;

    private int reason;
    private int playerId;
    private int result;
    private int unknown;

    @Override
    public int getRecordId() {
        return TYPE;
    }

    @Override
    public void parse(ByteBuffer inBuffer) {
        reason = inBuffer.getInt();
        playerId = inBuffer.get() & 0xFF;
        result = inBuffer.getInt();
        unknown = inBuffer.getInt();
    }

    @Override
    public ByteBuffer assembly(ByteBuffer outBuffer) {

        if(outBuffer == null)
            outBuffer = ByteBuffer.allocate(14).order(ByteOrder.LITTLE_ENDIAN);

        outBuffer.putInt(reason);
        outBuffer.put((byte) playerId);
        outBuffer.putInt(result);
        outBuffer.putInt(unknown);
        return outBuffer;
    }

    public int getReason() {
        return reason;
    }

    public LeaveRecord setReason(int reason) {
        this.reason = reason;
        return this;
    }

    public int getPlayerId() {
        return playerId;
    }

    public LeaveRecord setPlayerId(int playerId) {
        this.playerId = playerId;
        return this;
    }

    public int getResult() {
        return result;
    }

    public LeaveRecord setResult(int result) {
        this.result = result;
        return this;
    }

    public int getUnknown() {
        return unknown;
    }

    public LeaveRecord setUnknown(int unknown) {
        this.unknown = unknown;
        return this;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o)
            return true;
        if (o == null || getClass() != o.getClass())
            return false;
        LeaveRecord that = (LeaveRecord) o;
        return reason == that.reason && playerId == that.playerId && result == that.result && unknown == that.unknown;
    }

    @Override
    public int hashCode() {
        return Objects.hash(reason, playerId, result, unknown);
    }
}

File: src/main/kotlin/raft/war/binary/parser/w3g/record/LeaveRecord.kt
package raft.war.binary.parser.w3g.record

import java.nio.ByteBuffer
import java.util.*

class LeaveRecord : RecordBase {
    override var timestamp: Long = 0

    var reason: Int = 0
    var playerId: Int = 0
    var result: Int = 0
    var unknown: Int = 0

    override fun parse(inBuffer: ByteBuffer) {
        reason = inBuffer.getInt()
        playerId = inBuffer.get().toInt() and 0xFF
        result = inBuffer.getInt()
        unknown = inBuffer.getInt()
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val that = other as LeaveRecord
        return reason == that.reason && playerId == that.playerId && result == that.result && unknown == that.unknown
    }

    override fun hashCode(): Int {
        return Objects.hash(reason, playerId, result, unknown)
    }

    companion object {
        const val ID: Int = 0x17
    }

    override fun toString(): String = "☠️Time: $timestamp,\tLeaveRecord\n" +
            "\treason: $reason\n" +
            "\tplayerId: $playerId\n" +
            "\tresult: $result\n" +
            "\tunknown: $unknown\n"
}
